File: frontend/src/utils/voyageDuration.ts
/**
 * Voyage duration configuration and delivery window calculation utilities.
 * 
 * Trip durations are from first day of loading window to destination,
 * including 2-day laycan allowance.
 * 
 * Voyage durations are now fetched from the database via the discharge ports API.
 */

// Cache for voyage durations fetched from API
let voyageDurationsCache: Record<string, { suez: number | null, cape: number | null }> = {}
let cacheLoaded = false

// List of valid routes
export const CIF_ROUTES = ["SUEZ", "CAPE"]

// Single-route destinations (no route selection needed, same-month delivery allowed)
// These destinations only have a Suez route with short voyage duration
export const SINGLE_ROUTE_DESTINATIONS = ["Djibouti", "Keamari/Fotco"]

/**
 * Check if a destination is a single-route destination.
 * Single-route destinations don't require route selection and allow same-month delivery.
 */
export function isSingleRouteDestination(destination: string | undefined | null): boolean {
  if (!destination) return false
  return SINGLE_ROUTE_DESTINATIONS.includes(destination)
}

/**
 * Interface for discharge port data from API
 */
export interface DischargePort {
  id: number
  name: string
  restrictions?: string
  voyage_days_suez?: number
  voyage_days_cape?: number
  is_active: boolean
  sort_order: number
}

/**
 * Load voyage durations from the API into the cache.
 * This should be called once when the app initializes or when discharge ports are updated.
 */
export async function loadVoyageDurations(): Promise<void> {
  try {
    const response = await fetch('/api/discharge-ports')
    if (response.ok) {
      const ports: DischargePort[] = await response.json()
      voyageDurationsCache = {}
      ports.forEach(port => {
        voyageDurationsCache[port.name] = {
          suez: port.voyage_days_suez ?? null,
          cape: port.voyage_days_cape ?? null
        }
      })
      cacheLoaded = true
    }
  } catch (error) {
    console.error('Failed to load voyage durations from API:', error)
  }
}

/**
 * Get the list of valid CIF destinations from the cache.
 */
export function getCifDestinations(): string[] {
  return Object.keys(voyageDurationsCache)
}

/**
 * Get voyage duration in days for a destination and route.
 * Uses cached data from the database.
 * For single-route destinations, returns the available route duration regardless of route param.
 */
export function getVoyageDuration(destination: string, route: string): number | null {
  const portData = voyageDurationsCache[destination]
  if (!portData) return null

  // For single-route destinations, return the only available route (suez)
  if (isSingleRouteDestination(destination)) {
    return portData.suez
  }

  if (route && route.toUpperCase() === 'SUEZ') {
    return portData.suez
  } else if (route && route.toUpperCase() === 'CAPE') {
    return portData.cape
  }

  return null
}

/**
 * Set voyage durations directly (used when ports are fetched elsewhere).
 */
export function setVoyageDurations(ports: DischargePort[]): void {
  voyageDurationsCache = {}
  ports.forEach(port => {
    voyageDurationsCache[port.name] = {
      suez: port.voyage_days_suez ?? null,
      cape: port.voyage_days_cape ?? null
    }
  })
  cacheLoaded = true
}

/**
 * Check if voyage durations have been loaded.
 */
export function isVoyageDurationsLoaded(): boolean {
  return cacheLoaded
}

/**
 * Parse the first day from a loading window string.
 * 
 * Supports formats:
 * - "01-05/01" (DD-DD/MM)
 * - "1-5/1" (D-D/M)
 * - "15-20" (DD-DD, uses provided month/year)
 */
export function parseLoadingWindowStart(
  loadingWindow: string,
  month: number,
  year: number
): Date | null {
  if (!loadingWindow) return null
  
  loadingWindow = loadingWindow.trim()
  
  try {
    // Format: "DD-DD/MM" or "D-D/M"
    if (loadingWindow.includes("/")) {
      const parts = loadingWindow.split("/")
      const dayPart = parts[0]  // "01-05" or "1-5"
      const monthPart = parts[1]  // "01" or "1"
      
      // Get first day
      const firstDay = parseInt(dayPart.split("-")[0], 10)
      const parsedMonth = parseInt(monthPart, 10)
      
      // Determine year - if parsed month < reference month, it might be next year
      let parsedYear = year
      if (parsedMonth < month) {
        parsedYear = year + 1
      }
      
      return new Date(parsedYear, parsedMonth - 1, firstDay)
    }
    
    // Format: "DD-DD" (use reference month/year)
    if (loadingWindow.includes("-")) {
      const firstDay = parseInt(loadingWindow.split("-")[0], 10)
      return new Date(year, month - 1, firstDay)
    }
    
    // Single day
    const firstDay = parseInt(loadingWindow, 10)
    return new Date(year, month - 1, firstDay)
    
  } catch {
    return null
  }
}

/**
 * Get the last day of a month.
 */
function getLastDayOfMonth(year: number, month: number): number {
  // month is 0-indexed for Date, so month+1 with day 0 gives last day of month
  return new Date(year, month, 0).getDate()
}

/**
 * Calculate the ETA (Estimated Time of Arrival) date.
 * ETA = First day of loading window + voyage duration
 * 
 * Returns a formatted date string like "Mar 24" or null if cannot calculate.
 */
export function calculateETA(
  loadingWindow: string,
  destination: string,
  route: string,
  month: number,
  year: number
): string | null {
  // Get voyage duration
  const duration = getVoyageDuration(destination, route)
  if (duration === null) return null
  
  // Parse loading window start date
  const loadingStart = parseLoadingWindowStart(loadingWindow, month, year)
  if (!loadingStart) return null
  
  // Calculate ETA
  const eta = new Date(loadingStart)
  eta.setDate(eta.getDate() + duration)
  
  // Format as "Mon DD" (e.g., "Mar 24")
  const monthNames = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
  return `${monthNames[eta.getMonth()]} ${eta.getDate()}`
}

/**
 * Calculate the full ETA date object.
 * Returns the Date object or null if cannot calculate.
 */
export function calculateETADate(
  loadingWindow: string,
  destination: string,
  route: string,
  month: number,
  year: number
): Date | null {
  // Get voyage duration
  const duration = getVoyageDuration(destination, route)
  if (duration === null) return null
  
  // Parse loading window start date
  const loadingStart = parseLoadingWindowStart(loadingWindow, month, year)
  if (!loadingStart) return null
  
  // Calculate ETA
  const eta = new Date(loadingStart)
  eta.setDate(eta.getDate() + duration)
  
  return eta
}

/**
 * Calculate the delivery window based on loading window, destination, and route.
 * 
 * Formula logic:
 * - Delivery Start = First day of Loading + Trip Duration
 * - Delivery End = Delivery Start + 14 days
 * - If same month: "(StartDay-EndDay/Month)"
 * - If cross-month: Use last 15 days of delivery start month
 */
export function calculateDeliveryWindow(
  loadingWindow: string,
  destination: string,
  route: string,
  month: number,
  year: number
): string | null {
  // Get voyage duration
  const duration = getVoyageDuration(destination, route)
  if (duration === null) return null
  
  // Parse loading window start date
  const loadingStart = parseLoadingWindowStart(loadingWindow, month, year)
  if (!loadingStart) return null
  
  // Calculate delivery dates
  const deliveryStart = new Date(loadingStart)
  deliveryStart.setDate(deliveryStart.getDate() + duration)
  
  const deliveryEnd = new Date(loadingStart)
  deliveryEnd.setDate(deliveryEnd.getDate() + duration + 14)
  
  // Check if same month
  if (deliveryStart.getMonth() === deliveryEnd.getMonth()) {
    // Same month: (StartDay-EndDay/Month)
    const monthNum = deliveryStart.getMonth() + 1  // Convert to 1-indexed
    return `(${deliveryStart.getDate()}-${deliveryEnd.getDate()}/${monthNum})`
  } else {
    // Cross-month: Use last 15 days of delivery start month
    const deliveryMonth = deliveryStart.getMonth() + 1  // 1-indexed
    const lastDayOfMonth = getLastDayOfMonth(deliveryStart.getFullYear(), deliveryMonth)
    const adjustedStart = lastDayOfMonth - 14  // 15 days from end
    return `(${adjustedStart}-${lastDayOfMonth}/${deliveryMonth})`
  }
}

File: backend/app/utils/voyage_duration.py
"""
Voyage duration configuration and delivery window calculation utilities.

Trip durations are from first day of loading window to destination,
including 2-day laycan allowance.
"""

from datetime import date, timedelta
from calendar import monthrange
from typing import Optional, Tuple
from sqlalchemy.orm import Session

# List of valid routes
CIF_ROUTES = ["SUEZ", "CAPE"]

# Single-route destinations (no route selection needed, same-month delivery allowed)
# These destinations only have a Suez route with short voyage duration
SINGLE_ROUTE_DESTINATIONS = ["Djibouti", "Keamari/Fotco"]


def is_single_route_destination(destination: str) -> bool:
    """
    Check if a destination is a single-route destination.
    Single-route destinations don't require route selection and allow same-month delivery.

    Args:
        destination: Destination port name

    Returns:
        True if single-route destination, False otherwise
    """
    return destination in SINGLE_ROUTE_DESTINATIONS


def get_single_route_voyage_duration(destination: str, db: Session = None) -> Optional[int]:
    """
    Get voyage duration for a single-route destination.

    Args:
        destination: Destination port name
        db: SQLAlchemy database session (optional)

    Returns:
        Number of days for the voyage, or None if not found
    """
    if db is None:
        return None

    from app import models

    discharge_port = db.query(models.DischargePort).filter(
        models.DischargePort.name == destination
    ).first()

    if not discharge_port:
        return None

    # For single-route destinations, return the suez duration (the only available route)
    return discharge_port.voyage_days_suez


def get_voyage_duration(destination: str, route: str, db: Session = None) -> Optional[int]:
    """
    Get voyage duration in days for a destination and route.

    Fetches from database if db session provided, otherwise returns None.
    For single-route destinations, returns the available route duration regardless of route param.

    Args:
        destination: Destination port name (e.g., "Rotterdam", "Milford Haven")
        route: Route name ("SUEZ" or "CAPE") - ignored for single-route destinations
        db: SQLAlchemy database session (optional)

    Returns:
        Number of days for the voyage, or None if not found
    """
    if db is None:
        return None

    from app import models

    discharge_port = db.query(models.DischargePort).filter(
        models.DischargePort.name == destination
    ).first()

    if not discharge_port:
        return None

    # For single-route destinations, return the only available route (suez)
    if is_single_route_destination(destination):
        return discharge_port.voyage_days_suez

    if route and route.upper() == "SUEZ":
        return discharge_port.voyage_days_suez
    elif route and route.upper() == "CAPE":
        return discharge_port.voyage_days_cape

    return None


def get_voyage_duration_from_port(discharge_port, route: str) -> Optional[int]:
    """
    Get voyage duration from a DischargePort model instance.

    Args:
        discharge_port: DischargePort model instance
        route: Route name ("SUEZ" or "CAPE") - ignored for single-route destinations

    Returns:
        Number of days for the voyage, or None if not found
    """
    if not discharge_port:
        return None

    # For single-route destinations, return the only available route (suez)
    if is_single_route_destination(discharge_port.name):
        return discharge_port.voyage_days_suez

    if route and route.upper() == "SUEZ":
        return discharge_port.voyage_days_suez
    elif route and route.upper() == "CAPE":
        return discharge_port.voyage_days_cape

    return None


def parse_loading_window_start(loading_window: str, month: int, year: int) -> Optional[date]:
    """
    Parse the first day from a loading window string.
    
    Supports formats:
    - "01-05/01" (DD-DD/MM)
    - "1-5/1" (D-D/M)
    - "15-20" (DD-DD, uses provided month/year)
    
    Args:
        loading_window: Loading window string
        month: Reference month (1-12)
        year: Reference year
    
    Returns:
        First day of loading window as date, or None if parsing fails
    """
    if not loading_window:
        return None
    
    loading_window = loading_window.strip()
    
    try:
        # Format: "DD-DD/MM" or "D-D/M"
        if "/" in loading_window:
            parts = loading_window.split("/")
            day_part = parts[0]  # "01-05" or "1-5"
            month_part = parts[1]  # "01" or "1"
            
            # Get first day
            first_day = int(day_part.split("-")[0])
            parsed_month = int(month_part)
            
            # Determine year - if parsed month < reference month, it might be next year
            parsed_year = year
            if parsed_month < month:
                parsed_year = year + 1
            
            return date(parsed_year, parsed_month, first_day)
        
        # Format: "DD-DD" (use reference month/year)
        elif "-" in loading_window:
            first_day = int(loading_window.split("-")[0])
            return date(year, month, first_day)
        
        # Single day
        else:
            first_day = int(loading_window)
            return date(year, month, first_day)
            
    except (ValueError, IndexError):
        return None


def calculate_delivery_window(
    loading_window: str,
    destination: str,
    route: str,
    month: int,
    year: int,
    db: Session = None
) -> Optional[str]:
    """
    Calculate the delivery window based on loading window, destination, and route.
    
    Formula logic:
    - Delivery Start = First day of Loading + Trip Duration
    - Delivery End = Delivery Start + 14 days
    - If same month: "(StartDay-EndDay/Month)"
    - If cross-month: Use last 15 days of delivery start month
    
    Args:
        loading_window: Loading window string (e.g., "01-05/01")
        destination: Destination port name
        route: Route name ("SUEZ" or "CAPE")
        month: Reference month for loading window parsing
        year: Reference year for loading window parsing
        db: SQLAlchemy database session (optional)
    
    Returns:
        Delivery window string (e.g., "(9-23/2)"), or None if calculation fails
    """
    # Get voyage duration from database
    duration = get_voyage_duration(destination, route, db)
    if duration is None:
        return None
    
    # Parse loading window start date
    loading_start = parse_loading_window_start(loading_window, month, year)
    if loading_start is None:
        return None
    
    # Calculate delivery dates
    delivery_start = loading_start + timedelta(days=duration)
    delivery_end = loading_start + timedelta(days=duration + 14)
    
    # Check if same month
    if delivery_start.month == delivery_end.month:
        # Same month: (StartDay-EndDay/Month)
        return f"({delivery_start.day}-{delivery_end.day}/{delivery_start.month})"
    else:
        # Cross-month: Use last 15 days of delivery start month
        _, last_day_of_month = monthrange(delivery_start.year, delivery_start.month)
        adjusted_start = last_day_of_month - 14  # 15 days from end
        return f"({adjusted_start}-{last_day_of_month}/{delivery_start.month})"

